### DISCUSSION TITLE
Python MSAL acquire token by username and password

### DISCUSSION THREAD

**Discussion by GilPolo** [AzureAD/microsoft-authentication-library-for-python#594] [Q&A] [answered]
Hi,
I am using PowerBI-Developer-Samples/Python/Embed for your customers/AppOwnsData example to embed a Power BI report. Our Azure is tied to GoDaddy for authentication. The python code detects that the "account_type": "Federated", so then it sends a request to GoDaddy to get the token. I am receiving a 405 error code with "Method not Allowed" message. Appreciate any help, using simple user name password validation. But, should I user "ServicePrincipal" instead ?

url = https://sso.godaddy.com/v1/wss/trust/13/usernamemixed
url  = http://docs.oasis-open.org/ws-sx/ws-trust/200512/RST/Issue

Thanks,
Gilbert

**Comment by rayluo** [comment]
[Username password authentication has lots of limitation](https://github.com/AzureAD/microsoft-authentication-library-for-python/wiki/Username-Password-Authentication), and not recommended.

Have you tried the sample linked by the blue Desktop App icon in the ["scenarios map" here](https://msal-python.readthedocs.io/en/latest/)?

Service Principal is authenticating as the ID of that service principal, not an end user. If your application needs no user identity in the first place, Service Principal is a better choice.

**Reply by GilPolo** [reply]
Thank you, using ServicePrinciple and I am receiving a bearer token. However, the embed param call fails with a 401 error. https://api.powerbi.com/v1.0/myorg/groups/{workspace_id}/reports/{report_id}. If I manually login thru the Power BI Playground it works.

**Reply by rayluo** [reply, marked as answer]
When you "manually login thru the Power BI playground", were you using the same Service Principal or your own user account? Either way, now it sounds more like a "how to grant PowerBI access to a certain account" question to the PowerBI.
